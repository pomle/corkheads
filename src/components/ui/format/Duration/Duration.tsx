import React from "react";
import { Moment } from "moment";

interface DurationProps {
  from: Moment;
  until: Moment;
}

const UNITS: [number, string][] = [
  [60 * 60, "h"],
  [60, "min"]
];

const Duration: React.FC<DurationProps> = ({ from, until }) => {
  const parts: string[] = [];
  let rest = until.diff(from, "seconds");
  for (const [seconds, suffix] of UNITS) {
    const value = Math.floor(rest / seconds);
    rest -= value * seconds;
    if (value > 0) {
      parts.push(value + suffix);
    }
  }
  return <>{parts.join(" ")}</>;
};

export default Duration;
