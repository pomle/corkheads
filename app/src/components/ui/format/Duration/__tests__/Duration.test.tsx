import React from "react";
/* eslint import/no-extraneous-dependencies: ["error", {"peerDependencies": true}] */
import renderer from "react-test-renderer";
import moment from "moment";
import Duration from "../Duration";

describe("Duration component", () => {
  it("shows hour and minute", () => {
    const A = moment("2000-01-01T00:00:00Z");
    const B = A.clone()
      .add(3, "hours")
      .add(25, "minutes");
    const component = renderer.create(<Duration from={A} until={B} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("shows only minutes if under hour", () => {
    const A = moment("2000-01-01T00:00:00Z");
    const B = A.clone().add(59, "minutes");
    const component = renderer.create(<Duration from={A} until={B} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
