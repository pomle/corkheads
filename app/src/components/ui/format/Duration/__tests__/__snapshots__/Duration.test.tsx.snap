// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Duration component shows hour and minute 1`] = `"3h 25min"`;

exports[`Duration component shows only minutes if under hour 1`] = `"59min"`;
